src/lib/nnlib.jl

```diff
@@ -1,17 +1,9 @@
 using NNlib
 import NNlib: softmax, ∇softmax, logsoftmax, ∇logsoftmax, conv, depthwiseconv, ∇conv_data, ∇depthwiseconv_data, maxpool, meanpool, σ, relu
 
-Zygote.@adjoint function Base.broadcasted(::typeof(relu), x::Array{T}) where T<:Real
-    y = relu.(x)
-    return y, Δ -> begin
-        res = zero(Δ)
-        @inbounds for i in 1:length(res)
-            if y[i] > 0
-                res[i] = Δ[i]
-            end
-        end
-        (nothing, res)
-    end
+@adjoint function Base.Broadcast.broadcasted(::typeof(relu), xs::Array{T}) where T<:Real
```

Comment:
Could we relax this to make it work with `CuArray`s? Or is that covered elsewhere?

Reply:
Yes, this should definitely cover all array types. @findmyway there's a `Numeric` type in broadcast.jl that would be useful here, since it covers abstract arrays and scalars.

Reply:
I relaxed the constraint to `AbstractArray`. However, the improvement with `CuArray` is not that obvious compared to `Array`.

```julia

# before

#=
BenchmarkTools.Trial: 
  memory estimate:  6.52 KiB
  allocs estimate:  163
  --------------
  minimum time:     32.256 μs (0.00% GC)
  median time:      34.455 μs (0.00% GC)
  mean time:        39.182 μs (2.60% GC)
  maximum time:     38.206 ms (26.64% GC)
  --------------
  samples:          10000
  evals/sample:     1
=#

# after

#=
BenchmarkTools.Trial: 
  memory estimate:  4.50 KiB
  allocs estimate:  107
  --------------
  minimum time:     24.777 μs (0.00% GC)
  median time:      26.802 μs (0.00% GC)
  mean time:        30.936 μs (3.09% GC)
  maximum time:     36.959 ms (25.84% GC)
  --------------
  samples:          10000
  evals/sample:     1
=#
```

Reply:
OK, I'll switch to `Numeric`

Reply:
@MikeInnes Should I move the change to `broadcast.jl`, or better to keep it in the `nnlib.jl`?

Reply:
nnlib.jl is fine, given that it is nnlib-specific; just needs to follow the same template as the definitions in `broadcast.jl`.